# Student Performance Indicator
    Life cycle of the machine Learning Project
        1) Understanding the problem statement
        2) Data collection
        3) Data checks to perform
        4) Exploratory Data analysis
        5) Model Training
        6) Choosing the best model


# 1) Understanding the Problem statement
         This project is to understand how a students performance is graded or affected by Gender , Race ,Ethicinity , Parental Education
            lunch and Test Prep

# 2) The data is collected from kaggle  and it consists of 8 columns and 1000 rows

# 3) Package importing

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
from collections import OrderedDict

# read The data

In [5]:
df = pd.read_csv("Data\StudentsPerformance.csv")

# top 5 rows

In [6]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

# Shape of the Dataset

In [7]:
df.shape

(1000, 8)

# Data information
    

    gender : sex of students -> (Male/female)
    race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
    parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
    lunch : having lunch before test (standard or free/reduced)
    test preparation course : complete or not complete before test
    math score
    reading score
    writing score



# Data checks to perform
    
    Check Missing values
    Check Duplicates
    Check data type
    Check the number of unique values of each column
    Check statistics of data set
    Check various categories present in the different categorical column


# Check the missing values

In [8]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

There are no missing values

# Check for  duplicates

In [9]:
df.duplicated().sum()

0

There are no duplicate values

# Check the data types

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


# Checking the number of unique values in each column


In [14]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

# check the statistics of the data
    creating a custom fucntion for the descriptive statistics

In [16]:
def custom_summary(df): #parameters?? df = dataframe original data of mtcars
    
    result = []
    
    for col in df.columns:
        
        if df[col].dtype != 'O':
        
            stats = OrderedDict ({
                'Feature_Name' : col,
                'Count':df[col].count(),
                'Minimum':df[col].min(),
                'Quarter 1':df[col].quantile(0.25), 
                "Mean":df[col].mean(),
                'Median':df[col].median(),
                'Quarter 3':df[col].quantile(0.75),
                'Maximum':df[col].max(),
                "Variance":df[col].var(),
                'Standard Deviation':df[col].std(),
                "Kurtosis":df[col].kurt(),
                'Skewness':df[col].skew() ,
                'IQR':df[col].quantile(0.75) - df[col].quantile(0.25)
            })
            
            result.append(stats)
        
        
    result_df = pd.DataFrame(result)
    
    skewtype =[]
    
    for i in result_df['Skewness']:
        if i<=-1:
            skewtype.append('Highly Negatively Skewed')
        elif i<= -0.5:
            skewtype.append('Moderately Negatively Skewed')
        elif -0.5 < i < 0 :
            skewtype.append('Approx Normal Distribution (-ve)')
        elif 0 <= i < 0.5:
            skewtype.append('Approx Normal Distribution (+ve)')
        elif 0.5<= i < 1:
            skewtype.append('Moderately Positively Skewed')
        elif i >= 1:
            skewtype.append('Highly Positively Skewed')
    
    result_df['Skew_Type'] = skewtype
    
    k_type = []
    
    for i in result_df['Kurtosis']:
        if i <=  -1:
            k_type.append('Highly Platykurtic Curve')
        elif -1 < i <= -0.5:#-.9 to .9
            k_type.append('Moderately Platykurtic Curve')
        elif -0.5 < i <= 0.5: #-1,-2,-3...
            k_type.append('Mesokurtic Curve')
        elif 0.5<= i < 1:
            k_type.append('Moderately Leptokurtic Curve')
        elif i >= 1:
            k_type.append('Highly Leptokurtic Curve')
            
        
    
    result_df['Kurtosis_Type'] = k_type
    
    
    #Outlier detection
    

    Upper_limit = stats['Quarter 3'] + 1.5*stats['IQR']
    
    lower_limit = stats['Quarter 1'] -1.5*stats['IQR']
    
    if len([x for x in df[col] if (x < lower_limit) or (x > Upper_limit)]) > 0:
        
        outlier_comment = 'has outliers'
        outlier_percentage = len([x for x in df[col] if (x < lower_limit) or (x > Upper_limit)])/len(df)
    else:
        oulier_comment = 'no outliers'
        
        outlier_percentage = 0
        
    result_df['outlier_comment'] = outlier_comment
    
    result_df['outlier_percentage'] = outlier_percentage
   
    return result_df

In [17]:
custom_summary(df)

Feature_Name  Count  Minimum  Quarter 1    Mean  Median  Quarter 3  \
0     math score   1000        0      57.00  66.089    66.0       77.0   
1  reading score   1000       17      59.00  69.169    70.0       79.0   
2  writing score   1000       10      57.75  68.054    69.0       79.0   

   Maximum    Variance  Standard Deviation  Kurtosis  Skewness    IQR  \
0      100  229.918998           15.163080  0.274964 -0.278935  20.00   
1      100  213.165605           14.600192 -0.068265 -0.259105  20.00   
2      100  230.907992           15.195657 -0.033365 -0.289444  21.25   

                          Skew_Type     Kurtosis_Type outlier_comment  \
0  Approx Normal Distribution (-ve)  Mesokurtic Curve    has outliers   
1  Approx Normal Distribution (-ve)  Mesokurtic Curve    has outliers   
2  Approx Normal Distribution (-ve)  Mesokurtic Curve    has outliers   

   outlier_percentage  
0               0.005  
1               0.005  
2               0.005

1) The Data has very few outliers

2) From above description of numerical data, all means are very close to each other - between 66 and 68.05;

3) All standard deviations are also close - between 14.6 and 15.19;

4) While there is a minimum score 0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17



# Exploring the data

In [18]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [22]:
# printing out the unique values in each of the columns

print("Categories under 'Gender' Variable:        ",end="")
print(df['gender'].unique())

print("Categories under 'Parental Level Eduation' Variable:        ",end="")
print(df['parental level of education'].unique())

print("Categories under 'race_ethnicity' Variable:        ",end="")
print(df['race/ethnicity'].unique())

print("Categories under 'lunch' Variable:        ",end="")
print(df['lunch'].unique())

print("Categories under 'test_prep_course' Variable:        ",end="")
print(df['test preparation course'].unique())


Categories under 'Gender' Variable:        ['female' 'male']
Categories under 'Parental Level Eduation' Variable:        ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories under 'race_ethnicity' Variable:        ['group B' 'group C' 'group A' 'group D' 'group E']
Categories under 'lunch' Variable:        ['standard' 'free/reduced']
Categories under 'test_prep_course' Variable:        ['none' 'completed']


# defining categoricaL and Numerical Features for the dataset

In [23]:
numerical_features = [feature for feature in df.columns if df[feature].dtype !='O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']


print(" We have {} numerical features namely :{}".format(len(numerical_features),numerical_features))
print(" We have {} Categorical features namely :{}".format(len(categorical_features),categorical_features))

 We have 3 numerical features namely :['math score', 'reading score', 'writing score']
 We have 5 Categorical features namely :['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


# Adding extra columns for Total Score and "Average"

In [24]:
df['Total_Score'] = df['math score'] + df['reading score'] + df['writing score']

df["Average"] = df["Total_Score"]/3

df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   Total_Score    Average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

# lets Try drawing certain information from the data we have received so far about the performance of students



In [27]:
reading_full = df[df["reading score"]== 100]['Average'].count()
writing_full = df[df["writing score"]== 100]['Average'].count()
math_full = df[df["math score"]== 100]['Average'].count()



print(f"The Number of students with full marks in maths:{math_full}")
print(f"The Number of students with full marks in Reading:{reading_full}")
print(f"The Number of studets with full marks in Writing:{writing_full}")

The Number of students with full marks in maths:7
The Number of students with full marks in Reading:17
The Number of studets with full marks in Writing:14


now lets the see the lowest performance among the students
     lets consider the passing mark is 30


In [29]:
maths_low = df[df['math score'] <= 30]['Average'].count()
reading_low = df[df['reading score']<=30]["Average"].count()
writing_low = df[df['writing score']<=30]["Average"].count()


print(f"The number of students who failed to score good in maths :{maths_low}")
print(f"The Number of students who failed to score good in reading :{reading_low}")
print(f"The number of students who failed to score good marks in writing :{writing_low}")

The number of students who failed to score good in maths :16
The Number of students who failed to score good in reading :8
The number of students who failed to score good marks in writing :10


# The conclusions drawn from the follwing details
     1) Maths have been poorly performed by the students
     2) Results are pretty much good in reading